Description: temporary folder error handling
 Allow to left behind the temporary files generated by FireHOL in case of error.
 Backported from sanewall.
Author: Phil Whineray <phil@sanewall.org>
Last-Update: 2013-03-24
--- a/firehol.sh
+++ b/firehol.sh
@@ -307,6 +307,7 @@
 # WHY:
 # Even a CTRL-C will call this and we will not leave temp files.
 # Also, if a configuration file breaks, we will detect this too.
+FIREHOL_CLEAN_TMP=1
 FIREHOL_ACTIVATED_SUCCESSFULLY=0
 FIREHOL_SYSLOG_FACILITY="daemon"
 FIREHOL_NOTIFICATION_PROGRAM=""
@@ -337,7 +338,12 @@
 	fi
 	
 	# remove the temporary directory created for this session
-	test -d "${FIREHOL_DIR}" && ${RM_CMD} -rf "${FIREHOL_DIR}"
+	if [ ${FIREHOL_ACTIVATED_SUCCESSFULLY} -eq 0 -a ${FIREHOL_CLEAN_TMP} -eq 0 ]
+	then
+		echo "FireHOL: temporary files left in ${FIREHOL_DIR}"
+	else
+		test -d "${FIREHOL_DIR}" && ${RM_CMD} -rf "${FIREHOL_DIR}"
+	fi
 	
 	# syslog
 	local result=
@@ -7117,11 +7123,19 @@
 
 enable -n trap			# Disable the trap buildin shell command.
 enable -n exit			# Disable the exit buildin shell command.
-source ${FIREHOL_TMP} "$@"	# Run the configuration as a normal script.
+{ source ${FIREHOL_TMP} "$@"; }	# Run the configuration as a normal script.
+source_status=$?
 FIREHOL_LINEID="FIN"
 enable trap			# Enable the trap buildin shell command.
 enable exit			# Enable the exit buildin shell command.
 
+# There was a problem with the generated script - leave it behind
+if [ $source_status -ne 0 ]
+then
+  FIREHOL_CLEAN_TMP=0
+  exit $source_status
+fi
+
 close_cmd	|| ret=$[ret + 1]
 close_master	|| ret=$[ret + 1]
 
